src/js/core/methods/helpers/ethereumTypeDataConversions.js

```diff
@@ -0,0 +1,178 @@
+
+    const bigNumber = new BigNumber(number);
+
+    if (bigNumber.isGreaterThan(maxValue) || bigNumber.isLessThan(minValue)) {
+        throw new Error(`Overflow when trying to convert number ${number} into ${bytes} bytes`);
+    }
+
+    if (bigNumber.isPositive()) {
+        return bigNumber;
+    }
+    return bigNumber.minus(minValue).plus(minValue.negated());
+}
+
+// $FlowIssue bigint-unsupported, TODO: Update flow when bigint is supported
+function intToHex(
+    number: number | bigint | BigNumber | string,
+    bytes: number,
+    signed: boolean,
+): string {
+    let bigNumber = new BigNumber(number);
+    if (signed) {
+        bigNumber = twosCompliment(bigNumber, bytes);
+    }
+    const hex = bigNumber.toString(16);
+    const hexChars = bytes * 2;
+    if (hex.length > hexChars) {
+        throw new Error(`Overflow when trying to convert number ${number} into ${bytes} bytes`);
+    }
+    return hex.padStart(bytes * 2, '0');
+}
+
+/**
+ * Encodes the given primitive data to a big-endian hex string.
+ *
+ * @param typeName - Primitive Solidity data type (e.g. `uint16`)
+ * @param data - The actual data to convert.
+ * @returns Hex string of the data.
+ */
+export function encodeData(typeName: string, data: any): string | Uint8Array {
+    if (paramTypeBytes.test(typeName) || typeName === 'address') {
+        return stripHexPrefix(data);
+    }
+    if (typeName === 'string') {
+        return new TextEncoder().encode(data);
+    }
+    const numberMatch = paramTypeNumber.exec(typeName);
+    if (numberMatch) {
+        const [_, intType, bits] = numberMatch;
+        const bytes = Math.ceil(parseInt(bits, 10) / 8);
+        return intToHex(data, bytes, intType === 'int');
+    }
+    if (typeName === 'bool') {
+        return data ? '01' : '00';
+    }
+
+    // We should be receiving only atomic, non-array types
+    throw ERRORS.TypedError(
+        'Method_InvalidParameter',
+        `Unsupported data type for direct field encoding: ${typeName}`,
+    );
+}
+
+// these are simple types, so we can just do a string-match
+const paramTypesMap = {
+    string: Enum_EthereumDataType.STRING,
+    bool: Enum_EthereumDataType.BOOL,
+    address: Enum_EthereumDataType.ADDRESS,
+};
+
+/**
+ * Converts the given EIP-712 typename into a Protobuf package.
+ *
+ * @param typeName - The EIP-712 typename (e.g. `uint16` for simple types, `Example` for structs)
+ * @param types - Map of types, required for recursive (`struct`) types.
+ */
+export function getFieldType(
+    typeName: string,
+    types: EthereumSignTypedDataTypes,
+): EthereumFieldType {
+    const arrayMatch = paramTypeArray.exec(typeName);
+    if (arrayMatch) {
+        const [_, arrayItemTypeName, arraySize] = arrayMatch;
+        const entryType = getFieldType(arrayItemTypeName, types);
+        if (entryType.data_type === Enum_EthereumDataType.ARRAY) {
+            throw ERRORS.TypedError('Method_InvalidParameter', 'Nested arrays are not supported');
```

Comment:
you don't need to reproduce the failure from trezorlib :) host side can and should support encoding nested arrays. we can fix this on the firmware side independently.

Reply:
I'll need to double check if EIP-712 supports that this evening. If it doesn't, I'll update the error message to something like `EIP-712 doesn't support nested arrays`.

I believe solidity (Ethereum EVM programming language) doesn't support nested arrays, and that's where EIP-712 types come from.

Reply:
The spec doesn't say either way, so we'd need to look into implementations. Metamask's eth-sig-util does not have test vectors for nested arrays, so we could guess that they don't support it.

Reply:
I've removed this error and added some unit tests for type parsing in 36d1201398ffc43e507922e565c06b5add20ee27.

I couldn't find anywhere that explicitly says it's supported, but Metamask's `signTypedData_v4`'s implementation seems to support it in: https://github.com/MetaMask/eth-sig-util/blob/5f2259463990050606cd58b43e64e4bffcb715f4/src/sign-typed-data.ts#L202-L211

They use the last `[]` in the string to decode array types, e.g. `uint8[1][2]` would be an array of `uint8[1]` of length `2` (aka C-order/row-major, not Fortran-order).

Reply:
> They use the last [] in the string to decode array types

i recall this being explicitly specified _somewhere_, but can't find it now

Reply:
Ah, found it. https://github.com/ethereum/EIPs/pull/2190/files#diff-110e1894f4319a588e3e0d1ea324352f38af85215585c25029673e1ae5c6044dL140-R140

It's specified in an extension to EIP-712, which was unfortunately auto-closed after being stale for 9 months https://github.com/ethereum/EIPs/pull/2190

However, it does look like @cag and @MetaMask implemented those changes as part of `signTypedData_v4`: https://github.com/MetaMask/eth-sig-util/pull/54, and that's what most Dapps in the Ethereum eco-system uses.